index/score.go

```diff
@@ -245,23 +245,35 @@ func tfScore(k float64, b float64, L float64, f int) float64 {
 	return ((k + 1.0) * float64(f)) / (k*(1.0-b+b*L) + float64(f))
 }
 
+const importantTermBoost = 5
+const lowPriorityFilePenalty = 5
+
 // calculateTermFrequency computes the term frequency for the file match.
 // Notes:
 // - Filename matches count more than content matches. This mimics a common text search strategy to 'boost' matches on document titles.
 // - Symbol matches also count more than content matches, to reward matches on symbol definitions.
-func (p *contentProvider) calculateTermFrequency(cands []*candidateMatch) map[string]int {
+// - "Low priority" files like tests, generated files, etc. have their term frequency down-weighted, to prioritize matches from 'regular' files
+func (p *contentProvider) calculateTermFrequency(cands []*candidateMatch, lowPriority bool) map[string]int {
 	// Treat each candidate match as a term and compute the frequencies. For now, ignore case sensitivity and
 	// ignore whether the index is a word boundary.
 	termFreqs := map[string]int{}
 	for _, m := range cands {
 		term := string(m.substrLowered)
 		if m.fileName || p.matchesSymbol(m) {
-			termFreqs[term] += 5
+			termFreqs[term] += importantTermBoost
 		} else {
 			termFreqs[term]++
 		}
 	}
 
+	// If a file is a test, generated, etc., then down-weight its term frequency. The BM25F interpretation
+	// is that this data lives in a separate 'field' that is half the priority of regular content.
+	if lowPriority {
+		for term := range termFreqs {
+			termFreqs[term] = termFreqs[term] / lowPriorityFilePenalty
+		}
+	}
+
```

Comment:
I wonder whether it's more efficient to always have a penalty factor in the first loop and remove the second loop. So, for example, instead of `termFreqs[term]++` we have `termFreqs[term] = penalty * termFreqs[term]` where penalty is either 1 or `1/lorPriorityFilePenalty`.

Reply:
I'm not really seeing a way to do this while keeping `termFreqs` as an integer! Also do you really mean `termFreqs[term] = penalty * termFreqs[term]`? Isn't that a no-op if penalty is 1?

Let me know if I'm missing something and happy to fix in a follow-up.

Reply:
My point was to avoid a second loop. We anyway cast the termFreqs to float64 later on so that seems fine? Penalty would be 1 unless `lowPriority=true`. Again, I am not sure if that's an improvement. I would have to write a benchmark to be certain. I just like to avoid looping twice if once is enough. No need to follow up.